lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp

```diff
@@ -1540,12 +1607,398 @@ void amf_app_handle_initial_context_setup_rsp(
+  delete_packet_filter_t* pkt_filter = NULL;
+  for (int i = 0; i < qos_rule->no_of_pkt_filters; i++) {
+    pkt_filter = &delete_pkt_filter[i];
+    new_qos_rule_pkt_filter.pkt_filter_id = pkt_filter->identifier;
+    memcpy(qos_rule->new_qos_rule_pkt_filter, &new_qos_rule_pkt_filter,
+           1 * sizeof(NewQOSRulePktFilter));
+  }
+}
+
+static int pdu_session_resource_modification_t3591_handler(zloop_t* loop,
+                                                           int timer_id,
+                                                           void* arg) {
+  OAILOG_INFO(LOG_AMF_APP,
+              "T3591: pdu_session_resource_modification_t3591_handler\n");
+
+  amf_ue_ngap_id_t amf_ue_ngap_id = 0;
+  uint8_t pdu_session_id = 0;
+  ue_pdu_id_t uepdu_id;
+  std::shared_ptr<smf_context_t> smf_ctx;
+  char imsi[IMSI_BCD_DIGITS_MAX + 1];
+  int rc = 0;
+
+  if (!amf_pop_pdu_timer_arg(timer_id, &uepdu_id)) {
+    OAILOG_WARNING(LOG_AMF_APP,
+                   "T3591: Invalid Timer Id expiration, Timer Id: %u\n",
+                   timer_id);
+    OAILOG_FUNC_RETURN(LOG_NAS_AMF, RETURNok);
+  }
+
+  amf_ue_ngap_id = uepdu_id.ue_id;
+  pdu_session_id = uepdu_id.pdu_id;
+
+  ue_m5gmm_context_s* ue_context =
+      amf_ue_context_exists_amf_ue_ngap_id(amf_ue_ngap_id);
+
+  if (ue_context) {
+    IMSI64_TO_STRING(ue_context->amf_context.imsi64, imsi, 15);
+    smf_ctx = amf_get_smf_context_by_pdu_session_id(ue_context, pdu_session_id);
+
+    if (smf_ctx == NULL) {
+      OAILOG_ERROR(LOG_AMF_APP,
+                   "T3591:pdu session  not found for session_id = %u\n",
+                   pdu_session_id);
+      OAILOG_FUNC_RETURN(LOG_AMF_APP, rc);
+    }
+  } else {
+    OAILOG_ERROR(LOG_AMF_APP,
+                 "T3591: ue context not found for UE ID = " AMF_UE_NGAP_ID_FMT,
+                 amf_ue_ngap_id);
+    OAILOG_FUNC_RETURN(LOG_AMF_APP, rc);
+  }
+
+  OAILOG_WARNING(LOG_AMF_APP,
+                 "T3591: timer id: %ld expired for pdu_session_id: %d\n",
+                 smf_ctx->T3592.id, pdu_session_id);
+
+  smf_ctx->retransmission_count += 1;
+
+  OAILOG_ERROR(LOG_AMF_APP, "T3591: Incrementing retransmission_count to %d\n",
+               smf_ctx->retransmission_count);
+
+  if (smf_ctx->retransmission_count < PDU_SESS_MODFICATION_COUNTER_MAX) {
+    /* Send entity pdu session modification command
+     * message to the UE */
+
+    ue_pdu_id_t id = {amf_ue_ngap_id, pdu_session_id};
+    bstring nas_msg = bstrcpy(smf_ctx->session_message);
+    amf_app_handle_nas_dl_req(amf_ue_ngap_id, nas_msg, M5G_AS_SUCCESS);
+
+    smf_ctx->T3591.id = amf_pdu_start_timer(
+        PDU_SESSION_MODIFICATION_TIMER_MSECS, TIMER_REPEAT_ONCE,
+        pdu_session_resource_modification_t3591_handler, id);
+  } else {
+    /* Abort the registration procedure */
+    OAILOG_ERROR(
+        LOG_AMF_APP,
+        "T3591: Maximum retires:%d, for PDU_SESSION_MODIFICATION_REQUEST done "
+        "hence Abort the pdu sesssion modification procedure\n",
```

Comment:
PDU session abort is not complete

Reply:
Added Code to send negative response to sessiond